### PR TITLE
Split off predecessor tree's tree

### PR DESCRIPTION
**Pull request by nick-drozd**
Currently the predecessor tree's data structure is not particularly efficient. Separating the data structure for the predecessor tree itself will make it easier to improve.

### CODE REVIEW

File: validator/sawtooth_validator/execution/predecessor_tree.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def __repr__(self):
         return repr(retval)
 
 
-class PredecessorTree:
+class Tree:
```
Comment (vaporos) [inline validator/sawtooth_validator/execution/predecessor_tree.py]: What we want to end up with here is a generic RadixTrie implementation, so 'RadixTree' or 'RadixTrie' would be sensible names.

File: validator/sawtooth_validator/execution/predecessor_tree.py [hunk 2]
```diff
@@ -86,12 +86,17 @@ def get_nodes_above(self, address):
...
             yield node
 
+
+class PredecessorTree:
+    def __init__(self, token_size=2):
+        self._tree = Tree(token_size)
+
     def add_reader(self, address, reader):
-        node = self.insert(address)
+        node = self._tree.insert(address)
```
Comment (vaporos) [inline validator/sawtooth_validator/execution/predecessor_tree.py]: The radix tree should not allocate the stored data structure; instead, this should be "self._tree.insert(address, node_data)" where node could be a PredecessorTreeNode or any other python data structure.

Comment (nick-drozd) [reply validator/sawtooth_validator/execution/predecessor_tree.py]: fixed

File: validator/sawtooth_validator/execution/predecessor_tree.py (hunk 1 above)
Comment (nick-drozd) [reply validator/sawtooth_validator/execution/predecessor_tree.py]: The name can be changed when the data structure changes. Alternatively, the name can be left vague, and the specifics of the data structure can be changed without the possibility of rendering the name inaccurate.

File: validator/sawtooth_validator/execution/scheduler_parallel.py [hunk 3]
```diff
@@ -65,10 +65,10 @@ def __repr__(self):
         return repr(self._root)
 
     def _tokenize_address(self, address):
-        return [
+        return (
             address[i:i + self._token_size]
             for i in range(0, len(address), self._token_size)
-        ]
+        )
```
Comment (aludvik) [inline validator/sawtooth_validator/execution/scheduler_parallel.py]: This change should probably be in a separate commit.

File: validator/sawtooth_validator/execution/scheduler_parallel.py [hunk 4]
```diff
@@ -96,6 +96,15 @@ def get(self, address):
 
         return node
 
+    def get_nodes_above(self, address):
```
Comment (aludvik) [inline validator/sawtooth_validator/execution/scheduler_parallel.py]: A docstring would be helpful here, to clarify what `above` means

Comment (nick-drozd) [reply validator/sawtooth_validator/execution/scheduler_parallel.py]: Renamed to `get_ancestors` and docstring added

File: validator/sawtooth_validator/execution/scheduler_parallel.py (hunk 3 above)
Comment (nick-drozd) [reply validator/sawtooth_validator/execution/scheduler_parallel.py]: I don't think it's worth a full commit, but I added a comment about it in the `get_ancestors` commit.

File: validator/sawtooth_validator/execution/scheduler_parallel.py [hunk 5]
```diff
@@ -111,14 +105,44 @@ def get_ancestors(self, address):
...
+        self._tree = Tree(token_size=token_size)
+
+    def __repr__(self):
+        return repr(self._tree)
+
+    def _insert(self, address):
+        node = self._tree.insert(address)
+
+        if node.data is None:
+            node.data = Predecessors()
```
Comment (aludvik) [inline validator/sawtooth_validator/execution/scheduler_parallel.py]: Can you just pass the new instance of Predecessor to `self._tree.insert`?

Comment (nick-drozd) [reply validator/sawtooth_validator/execution/scheduler_parallel.py]: No. The logic of `insert` (maybe it needs a different name) is to get the node at address if it exists, and create it if it doesn't. In the former case, we don't necessarily want to overwrite the existing data.

**Comment by peterschwarz** [comment]
Is there anyway to add a unit test for the `Tree` class by itself?  I'm having trouble understanding what it is going to look like in use outside of the PredecessorTree.

File: validator/sawtooth_validator/execution/scheduler_parallel.py [hunk 6]
```diff
@@ -36,67 +35,114 @@
...
-        if self.children:
-            retval['children'] = \
-                {k: literal_eval(repr(v)) for k, v in self.children.items()}
 
-        return repr(retval)
-
-
-class PredecessorTree:
-    def __init__(self, token_size=2):
+class Tree:
```
Comment (jsmitchell) [inline validator/sawtooth_validator/execution/scheduler_parallel.py]: Calling this a Tree is confusing, since its interface doesn't follow expected Tree semantics. Call this a RadixTree or AddressedTree or something better.
